crates/vm2/src/predication.rs

```diff
@@ -4,43 +4,46 @@ const GT_BIT: u8 = 1 << 2;
 const ALWAYS_BIT: u8 = 1 << 3;
 
 #[derive(Debug, Clone, PartialEq)]
-pub struct Flags(u8);
+pub(crate) struct Flags(u8);
 
 impl Flags {
-    pub fn new(lt_of: bool, eq: bool, gt: bool) -> Self {
-        Flags(lt_of as u8 | ((eq as u8) << 1) | ((gt as u8) << 2) | ALWAYS_BIT)
+    pub(crate) fn new(lt_of: bool, eq: bool, gt: bool) -> Self {
+        Flags(u8::from(lt_of) | (u8::from(eq) << 1) | (u8::from(gt) << 2) | ALWAYS_BIT)
     }
 }
 
-/// Predicate encoded so that comparing it to flags is efficient
-#[derive(Copy, Clone, Debug, Hash)]
+/// Predicate for an instruction. Encoded so that comparing it to flags is efficient.
+#[derive(Copy, Clone, Debug, Default, Hash)]
 #[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
 #[repr(u8)]
 pub enum Predicate {
+    /// Always execute the associated instruction.
+    #[default]
     Always = ALWAYS_BIT,
+    /// Execute the associated instruction if the "greater than" execution flag is set.
     IfGT = GT_BIT,
+    /// Execute the associated instruction if the "equal" execution flag is set.
     IfEQ = EQ_BIT,
+    /// Execute the associated instruction if the "less than" execution flag is set.
     IfLT = LT_BIT,
+    /// Execute the associated instruction if either of "greater than" or "equal" execution flags are set.
     IfGE = GT_BIT | EQ_BIT,
+    /// Execute the associated instruction if either of "less than" or "equal" execution flags are set.
     IfLE = LT_BIT | EQ_BIT,
+    /// Execute the associated instruction if the "equal" execution flag is not set.
     IfNotEQ = EQ_BIT << 4 | ALWAYS_BIT,
```

Comment:
This is cute but just having `NOT_EQ_BIT = 1 << 4` and handling the special case might be faster and clearer.

Reply:
Sorry, I don't understand this suggestion; e.g., what is the special case here? You can implement it separately 🙃

Reply:
(not important) All the other cases are handled by `predicate & flags != 0` but that can't handle IfNotEq, so that one is handled by having the next four bits mean flags that we don't want.